File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/CBuilder/RTTIInspectorDemo/RTTIInspectorDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "RTTIInspectorDemoMain.h"
#include "RTTIInspectorDemoData.h"
#include "RTTIInspectorDemoVGEditor.h"
#include "cxExportVGLink.hpp"
#include "cxImageComboBox.hpp"
#include <assert.h>
#include <ImgList.hpp>
#include "RTTIInspectorDemoPropEditors.h"
#include "dxCore.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxDBVGrid"
#pragma link "cxInplaceContainer"
#pragma link "cxOI"
#pragma link "cxVGrid"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TRTTIInspectorDemoMainForm *RTTIInspectorDemoMainForm;
//---------------------------------------------------------------------------

PTypeInfo cxGetTypeInfo(PTypeInfo ATypeInfo, String APropName)
{
  PPTypeInfo Temp;
  Temp = GetPropInfo(ATypeInfo, APropName)->PropType;
  assert(Temp != NULL);
  return *Temp;
}

__fastcall TRTTIInspectorDemoMainForm::TRTTIInspectorDemoMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
  PTypeInfo AInfo = cxGetTypeInfo(__typeinfo(TcxCustomEditorRowProperties),
    "EditPropertiesClassName");
  cxRegisterPropertyEditor(AInfo, __classid(TcxCustomEditorRowProperties),
    "EditPropertiesClassName", NULL);
  AInfo = cxGetTypeInfo(__typeinfo(TcxCustomEditorRowProperties), "ImageIndex");
  cxRegisterPropertyEditor(AInfo , __classid(TcxCustomEditorRowProperties),
    "ImageIndex", __classid(TcxImageIndexProperty));
  cxRegisterEditPropertiesClass(__classid(TcxImageIndexProperty),
    __classid(TcxImageComboBoxProperties));
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::actVGEditExecute(TObject *Sender)
{
  GetVerticalGridEditor(cxDBVerticalGrid, FSelectedObjectEvent)->Show();
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::FormCreate(TObject *Sender)
{
  TDemoBasicMainForm::FormCreate(Sender);
  FSelectedObjectEvent = ObjectSelected;
  ObjectSelected(cxDBVerticalGrid);
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::cxRTTIInspectorPropertyChanged(TObject *Sender)
{
  if (((TcxEditorRow*)((TcxRTTIInspector*)Sender)->FocusedRow)->Properties->Caption == "Name")
    GetVerticalGridEditor(cxDBVerticalGrid, FSelectedObjectEvent)->DoItemsModified();
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::cxVerticalGridClick(TObject *Sender)
{
  ObjectSelected(Sender);
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::actExportVGToHTMLExecute(TObject *Sender)
{
  cxExportVGToHTML("temp.html", cxDBVerticalGrid, false, 8, "html");
  dxShellExecute("temp.html");
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::actExportVGToExcelExecute(TObject *Sender)
{
  cxExportVGToXLSX("temp.xlsx", cxDBVerticalGrid, false, true, 8, "xlsx");
  ShellExecute(Handle, "OPEN", "temp.xlsx", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::cxDBVerticalGridLayoutChanged(TObject *Sender)
{
  GetVerticalGridEditor(cxDBVerticalGrid, FSelectedObjectEvent)->DoItemsModified();
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::actShowCustomizeExecute(TObject *Sender)
{
  cxDBVerticalGrid->Customizing->Visible = true;
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::actAboutExecute(TObject *Sender)
{
  ShowAbout(false, true);
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::ObjectSelected(TObject *Sender)
{
  cxRTTIInspector->InspectedObject = (TPersistent*)Sender;
  UpdateLabel((TComponent*)Sender);
}
//---------------------------------------------------------------------------

void __fastcall TRTTIInspectorDemoMainForm::UpdateLabel(TComponent *AObject)
{
  if (AObject != NULL) {
    lbInspectedObject->Caption = AObject->Name + ": " + AObject->ClassName();
  } else {
    lbInspectedObject->Caption = "<not selected>";
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/Delphi/RTTIInspectorDemo/RTTIInspectorDemoMain.pas
unit RTTIInspectorDemoMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, DemoBasicMain, cxLookAndFeels, ActnList, ImgList, Menus,
  StdCtrls, ComCtrls, cxVGrid, cxControls, cxInplaceContainer, cxOI,
  cxClasses, cxStyles, RTTIInspectorDemoVGEditor, cxGraphics, cxEdit,
  cxCalc, cxExportVGLink, cxDBVGrid, ExtCtrls, cxLookAndFeelPainters,
  cxBlobEdit, cxMemo;

type
  TRTTIInspectorDemoMainForm = class(TDemoBasicMainForm)
    actVGEdit: TAction;
    VerticalEdit1: TMenuItem;
    PopupMenu: TPopupMenu;
    VerticalGridEdit1: TMenuItem;
    VerticalGridLayoutEditor1: TMenuItem;
    cxStyleRepository1: TcxStyleRepository;
    actExportVGToHTML: TAction;
    actExportVGToExcel: TAction;
    N1: TMenuItem;
    ExportToExcel1: TMenuItem;
    ExportToHTML1: TMenuItem;
    N2: TMenuItem;
    cxVerticalGridStyleSheetDevExpress: TcxVerticalGridStyleSheet;
    cxStyle1: TcxStyle;
    cxStyle2: TcxStyle;
    cxStyle3: TcxStyle;
    cxStyle4: TcxStyle;
    cxStyle5: TcxStyle;
    cxStyle6: TcxStyle;
    cxStyle7: TcxStyle;
    cxDBVerticalGrid: TcxDBVerticalGrid;
    ImageList: TImageList;
    Splitter1: TSplitter;
    cxDBVerticalGridID: TcxDBEditorRow;
    cxDBVerticalGridCustomerID: TcxDBEditorRow;
    cxDBVerticalGridProductID: TcxDBEditorRow;
    cxDBVerticalGridPurchaseDate: TcxDBEditorRow;
    cxDBVerticalGridPaymentType: TcxDBEditorRow;
    cxDBVerticalGridTime: TcxDBEditorRow;
    cxDBVerticalGridPaymentAmount: TcxDBEditorRow;
    cxDBVerticalGridDescription: TcxDBEditorRow;
    cxDBVerticalGridQuantity: TcxDBEditorRow;
    cxDBVerticalGridCustomerEmail: TcxDBEditorRow;
    actShowCustomize: TAction;
    Aboutthisdemo1: TMenuItem;
    Panel1: TPanel;
    cxRTTIInspector: TcxRTTIInspector;
    lbInspectedObject: TLabel;
    cxDBVerticalGridDCar: TcxDBEditorRow;
    procedure actVGEditExecute(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure cxRTTIInspectorPropertyChanged(Sender: TObject);
    procedure cxVerticalGridClick(Sender: TObject);
    procedure actExportVGToHTMLExecute(Sender: TObject);
    procedure actExportVGToExcelExecute(Sender: TObject);
    procedure cxDBVerticalGridLayoutChanged(Sender: TObject);
    procedure actShowCustomizeExecute(Sender: TObject);
  private
    FSelectedObjectEvent: TNotifyEvent;
    procedure ObjectSelected(Sender: TObject);
    procedure UpdateLabel(AObject: TComponent);
  end;

var
  RTTIInspectorDemoMainForm: TRTTIInspectorDemoMainForm;

implementation

uses DemoBasicAbout, DemoRating, ShellAPI, RTTIInspectorDemoData, dxCore;

{$R *.dfm}

procedure TRTTIInspectorDemoMainForm.actVGEditExecute(Sender: TObject);
begin
  GetVerticalGridEditor(cxDBVerticalGrid, FSelectedObjectEvent).Show;
end;

procedure TRTTIInspectorDemoMainForm.FormCreate(Sender: TObject);
begin
  inherited;
  FSelectedObjectEvent := ObjectSelected;
  ObjectSelected(cxDBVerticalGrid);
end;

procedure TRTTIInspectorDemoMainForm.ObjectSelected(Sender: TObject);
begin
  cxRTTIInspector.InspectedObject := TPersistent(Sender);
  UpdateLabel(TComponent(Sender));
end;

procedure TRTTIInspectorDemoMainForm.cxRTTIInspectorPropertyChanged(
  Sender: TObject);
begin
  if TcxEditorRow(TcxRTTIInspector(Sender).FocusedRow).Properties.Caption = 'Name' then
    GetVerticalGridEditor(cxDBVerticalGrid, FSelectedObjectEvent).DoItemsModified;
end;

procedure TRTTIInspectorDemoMainForm.cxVerticalGridClick(Sender: TObject);
begin
  ObjectSelected(Sender);
end;

procedure TRTTIInspectorDemoMainForm.actExportVGToHTMLExecute(Sender: TObject);
begin
  cxExportVGToHTML('temp.html', cxDBVerticalGrid, False);
  dxShellExecute('temp.html');
end;

procedure TRTTIInspectorDemoMainForm.actExportVGToExcelExecute(
  Sender: TObject);
begin
  cxExportVGToXLSX('temp.xlsx', cxDBVerticalGrid, False);
  ShellExecute(Handle, 'OPEN', 'temp.xlsx', nil, nil, SW_SHOWMAXIMIZED);
end;

procedure TRTTIInspectorDemoMainForm.cxDBVerticalGridLayoutChanged(
  Sender: TObject);
begin
  GetVerticalGridEditor(cxDBVerticalGrid, FSelectedObjectEvent).DoItemsModified;
end;

procedure TRTTIInspectorDemoMainForm.actShowCustomizeExecute(Sender: TObject);
begin
  cxDBVerticalGrid.Customizing.Visible := True;
end;

procedure TRTTIInspectorDemoMainForm.UpdateLabel(AObject: TComponent);
begin
  if AObject <> nil then
    lbInspectedObject.Caption := Format('%s: %s', [AObject.Name, AObject.ClassName])
  else
    lbInspectedObject.Caption := '<not selected>';
end;

initialization
  cxRegisterPropertyEditor(TypeInfo(string), TcxCustomEditorRowProperties, 'EditPropertiesClassName', nil);
end.
